Qualisys/QTM-Unity-Realtime-Streaming/Helpers/RTStreamThread.cs

```diff
@@ -106,30 +106,44 @@ public bool Update()
             {
                 ReaderThreadState.CopyFrom(writerThreadState);
             }
-            if (ReaderThreadState.connectionState == RTConnectionState.Disconnected)
-            {
-                Debug.Log(ReaderThreadState.errorString);
-            }
             return ReaderThreadState.connectionState != RTConnectionState.Disconnected;
         }
 
         void WriterThreadFunction() 
         {
             try
             {
-                using (var rtProtocol = new RTProtocol(LOWEST_SUPPORTED_UNITY_MAJOR_VERSION, LOWEST_SUPPORTED_UNITY_MINOR_VERSION))
+                using (var rtProtocol = new RTProtocol(RT_LOWEST_SUPPORTED_VERSION_MAJOR, RT_LOWEST_SUPPORTED_VERSION_MINOR))
                 {
-                    if (!rtProtocol.Connect(IpAddress, udpPort, RTProtocol.Constants.MAJOR_VERSION, RTProtocol.Constants.MINOR_VERSION))
+                    if (!rtProtocol.Connect(IpAddress, udpPort, RT_LOWEST_SUPPORTED_VERSION_MAJOR, RT_LOWEST_SUPPORTED_VERSION_MINOR)) 
                     {
-                        if (!rtProtocol.Connect(IpAddress, udpPort, LOWEST_SUPPORTED_UNITY_MAJOR_VERSION, LOWEST_SUPPORTED_UNITY_MINOR_VERSION))
+                        throw new WriterThreadException("Error Creating Connection to server" + rtProtocol.GetErrorString());
+                    }
+
+                    RtProtocolVersion version = new RtProtocolVersion(RTProtocol.Constants.MAJOR_VERSION, RTProtocol.Constants.MINOR_VERSION);
+                    
+                    //Upgrade protocol version
+                    for (; version.minor >= RT_LOWEST_SUPPORTED_VERSION_MINOR; --version.minor)
+                    {
+                        lock (syncLock)
                         {
-                            throw new WriterThreadException("Error Creating Connection to server" + rtProtocol.GetErrorString());
+                            writerThreadState.rtProtocolVersion.CopyFrom(version);
                         }
+                        string response;
+                        if (rtProtocol.SetVersion(version.major, version.minor, out response))
+                        {
+                            break;
+                        }
+                    }
+
+                    if (version.minor < RT_LOWEST_SUPPORTED_VERSION_MINOR)
+                    {
+                        throw new WriterThreadException("Failed to negotiate RT Protocol version with QTM");

```

Comment:
Maybe good to clarify to user what version is the lowest supported here aswell?

Reply:
If SetVersion fails until we reach v 1.12. I think it will be likely that a future QTM version has broken bakwards compability or that a user is having a really old version of QTM. That's why the error message is a bit generic.

Reply:
But if we get to this exception. Our version.minor is lower tha RT_LOWEST_SUPPORTED_VERSION_MINOR. Id say this should be communicated to the user. And if we can get here because of other fails, then I think we should try to catch these separately.

Reply:
Having more detailed errors is a good thing. However with the limited information we get from the RTClientSDK.NET it's abit more complicated to know which is the highest supported rt protocol version of QTM. It would involve complicating the negotiation process abit. I think it's a better addition for RTClientSDK.NET or even QTM

This exceptions happens when the RTClient is in the Connecting phase so in a Unity/User code perspective this is a connection error.